plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -7,10 +7,17 @@
 {-# LANGUAGE OverloadedLabels      #-}
 {-# LANGUAGE OverloadedStrings     #-}
 {-# LANGUAGE PackageImports        #-}
+{-# LANGUAGE PatternSynonyms       #-}
 {-# LANGUAGE ScopedTypeVariables   #-}
 {-# LANGUAGE TypeFamilies          #-}
 {-# OPTIONS_GHC -Wno-orphans   #-}
 
+#ifdef HLINT_ON_GHC_LIB
+#define MIN_GHC_API_VERSION(x,y,z) MIN_VERSION_ghc_lib(x,y,z)
+#else
+#define MIN_GHC_API_VERSION(x,y,z) MIN_VERSION_ghc(x,y,z)
+#endif
+
```

Comment:
Is it possible to drop this and simply assume that hlint is always on ghc-lib?

Reply:
Probably. I just followed the pattern from other parts of the code.

Reply:
just in case hlint is on ghc-lib except for the last ghc version (9.0 with this pr)

Reply:
I think some of the complexity is needed, since we need to do different things depending on what libraries hlint is using. This depends on both the version of hlint and the version of ghc and what flags hlint has set.

Reply:
I guess we could force hlint to always use ghc-lib, but that might be a bit redundant when the ghc-lib version is the same as the ghc version.

Reply:
I would prioritise the ease of maintaining here and drop support for hlint-on-ghc

Reply:
The reason to have two code paths (iirc from a discussion with @ndmitchell i dont find right now) is we have three components insteracting:
* ghcide: uses raw ghc
* hlint: by default uses raw ghc for last version and ghc-lib for previous ones
* apply-refact: uses raw ghcs

So we have a code path for the last ghc version which can get the parsed module from ghcide and use it directly for hlint and apply-refact. It is more precise and performant.
For ghc-lib we have to parse the file manually to give it to hlint and then write to a temp dir on disk the contents to call to apply-refact, which parses the file again, etc.

Reply:
The  discussion was here:
https://github.com/haskell/haskell-language-server/issues/32#issuecomment-637458673
and
https://github.com/haskell/haskell-language-server/issues/32#issuecomment-637587961

Reply:
Thanks for the clear explanation. I still think that supporting hlint-on-ghc is suicidal, given that hlint only supports 1 ghc version whereas we need to support a much wider range. Maintainability trumps performance. But this is just my opinion

Reply:
Not only performance, quoting the mentioned discussion:

> Better to share the AST, and likely to reduce the potential bugs (as things like the extensions will be precisely known), but otherwise it degrades reasonably.

But maybe we could reconsider the decision @ndmitchell?

Reply:
otoh once we have full ghc 9 support and drop ggc 8.6 we will only use one version of hlint, 3.3

Reply:
But we will need to maintain this hlint duality for the foreseeable future, which is not free. Anyways, we can always remove it later, resolving this.